### DISCUSSION TITLE
hdr tonemapping, how to configure?

### DISCUSSION THREAD

**Discussion by geextahslex** [mpv-player/mpv#15394] [Q&A]
Hi, I have a normal SDR TV and with mpv default settings I have noticed crushed blacks/artifacting in dark areas. So you loose detail and get some nasty color banding. This gets solved when you raise the brightness by 2-4 or switch `target-trc` from the default `pq` to `srgb` or `bt.1886` but this messes up dark scenes contrast and saturation.
I'm not sure what to do at this point. You either get one (consistent contrast, saturation but crushed blacks) or the other (fine blacks but messed up saturation and contrast)

You have to watch these screenshots at fullscreen because here in the preview they all look broken/have banding

Thank you :)

sample to the first scene
[Vikings.Valhalla.zip](https://github.com/user-attachments/files/17936220/Vikings.Valhalla.mkv.zip)

corners and her dress, hair
![crushed blacks artifacts](https://github.com/user-attachments/assets/586dc938-e883-4a85-a203-f3197e7560ef)
![crushed blacks artifacts bright2](https://github.com/user-attachments/assets/0511596c-ebc0-4798-a229-1636ca4c1114)

in the middle of the arrow where it goes from shadow to highlight
![shadows bending](https://github.com/user-attachments/assets/f8245772-b088-49cb-bf25-d33c4953ddd4)
![shadows bending brightness plus2](https://github.com/user-attachments/assets/d591dd26-824c-439c-9d81-5d7dc39cc81d)

marked with yellow, crushed red color
![artifacts](https://github.com/user-attachments/assets/0f212563-cb8b-45c6-bcdf-06b1131b10ab)
![artifacts no](https://github.com/user-attachments/assets/9491c3a8-69b5-4a57-8c8a-0eb93847da64)

**Comment by stonix015** [comment]
Hey, i might be able to help you here.

1) you stated that you switched the target-trc "from pq to srgb or bt.1886" ... i think all of those three are wrong, maybe you need to configure that to be gamma2.2 ... Because most of the displays are gamma2.2

2) Tone-Mapping is a thing that strongly depends on a personal opinion.
All of the HDR tone-mappers (BT-curves, spline) try to preserve your specified display brightness (e.g. 100, 150 or 200 nits, auto = 203), but then press anything between 200 and 1000 nits into the last 20 nits of your display, which makes bright scenes look very overexposed.

I ended up using the linear-tone-mapping curve, as it preserves the saturation and detail in bright scenes, but adapts to scenes that are just dark. Additionally, i set the hdr-peak-percentile to 99,8 ... In that case 0,2% of the bright details can be clipped out, but this declines that your mapping will pump around at dark scenes, where e.g. a candle is glimming or something like that.
0,2% is relatively a table lamp, which is 3-4 meters away from the camera in the scene. Difficult to explain, but it looked the best for me. 

Regarding the blacks:
It helped me a lot to set the target-contrast to values between 1000 and 5000 ..... 

The following tutorial depends if you have a device, where the brightness-lever adjusts the background light **(1)** (monitor, TV) or where the brightness-lever adjusts the panel-offset (projectors with static light source, e.g. UHP lamp, **(2)**):
First you could grab a very dark scene (are you able to get the movie "Bodies-Bodies-Bodies" ? Then 
start the movie with: mpv Bodies.mkv --start=1:21:38 --pause
This scene is absolutely georgios to set the black levels of the display.)
**1**
If you have got a display/monitor you should set the target-contrast until you see the details coming in, but without producing a greyish look. The scene should be absolutely dark at the borders (RGB-values = 0) 
**2**
If you have a projecetor, you should set the target-contrast to 2500 or 5000 statically and then adjust the brightness-lever of your projector until the "grey-borders" disappear. Contrahery, you can set the target-contrast=inf but then set the brightness-lever of your projector to 0 ..... In my case, the contrast=2500, but projector-lever to -3, really helped me getting details in the dark up. The picture got some serious details in the dark now, but without losing saturation and contrast in bright scenes.
Additionally, my projector got a gamma-setting curve "A" (its a JVC RS500) ... which really helped raising details in the dark.

**Comment by stonix015** [comment]
My next thought belongs into a second comment i guess. To keep things in order here.

I also had troubles with blacks, and i recognized that when setting everything like explained above, some movies got a really decent greyish look, while others are perfect. Some movies won't go into zero on the blacks, worse examples are Inception or Prey.

I was demanding for an adaptive black-level feature that measures the darkest pixels in the scene and adapts the tone-mapping curve so that pixels appear as zero:
https://github.com/haasn/libplacebo/discussions/303 (the third comment is the final solution that i programmed. Maybe haasn will response to it and maybe this feature can be implemented. At the moment, it is a workaround, you might read the thread.)

Ive attached the resulting compiled libplacebo-git version as an arch linux package:
[libplacebo-git_raw.zip](https://github.com/user-attachments/files/17974756/libplacebo-git_raw.zip)

My settings with this libplacebo version are:

```
#video driver
vo=gpu-next
gpu-api=opengl
fullscreen=yes
#drm-format=xrgb2101010
#drm-mode=3840x2160@30

#Audio
audio-channels=fl-fr-lfe-bl-br
ao=pulse

#Performance
cache=yes
demuxer-max-bytes=256MiB
hwdec=auto
dither=no
scale=bilinear
dscale=bilinear
cscale=bilinear
vd-lavc-threads=24
allow-delayed-peak-detect=yes

#syncing
video-sync=display-resample
video-sync-max-video-change=0.01
interpolation=yes

#General_Pic
target-prim = dci-p3
#tone-mapping-visualize = yes

[HDR]
profile-cond=p["video-params/primaries"] == "bt.2020"
vf = format:dolbyvision=no:hdr10plus=no
hdr-compute-peak=yes
hdr-peak-percentile=99.8
hdr-peak-decay-rate=20
hdr-contrast-recovery=0.5
target-peak=100 # NOTE! this doesnt have to do with my true screen brightness. Its just the brightest option, where the picture does not wildly pump around. Most of the movies have some wider HDR-ares that are around 70-100 nits.
tone-mapping =spline # NOTE! This works out as a linear-type tone-mapping with my custom libplacebo-version
tone-mapping-param = 0
target-trc =gamma2.2
target-contrast =2500

[SDR]
profile-cond=p["video-params/primaries"] and p["video-params/primaries"] ~= "bt.2020"
inverse-tone-mapping=yes
target-peak=203
tone-mapping = auto
target-trc = auto
target-contrast = auto
```

**Comment by geextahslex** [comment]
Okay thank you, I will look into that when I'm again in HDR mode, but for now after 2 weeks full of comparisons I am burned out xD
So far this worked for me:

So after further testing I ended up using [hdr-toys](https://github.com/natural-harmonia-gropius/hdr-toys) for all of the "color space" stuff (gamut-mapping and transfer function) and mpv only does the tone-mapping with 2094-10. 

On a SDR TV this gives you the best result. No clipped highlights, no crushed blacks, no washed out colors. Also good performance on crappy hardware :)
```
target-prim=bt.2020
target-trc=pq
tone-mapping=st2094-10
gamut-mapping-mode=clip (clip because hdr-toys handles this)
target-peak=100
hdr-compute-peak=yes
allow-delayed-peak-detect

glsl-shader=~~/shaders/hdr-toys/transfer-function/pq_inv.glsl
glsl-shader=~~/shaders/hdr-toys/utils/black_point_compensation.glsl
glsl-shader=~~/shaders/hdr-toys/gamut-mapping/jedypod.glsl
glsl-shader=~~/shaders/hdr-toys/transfer-function/bt1886.glsl
glsl-shader-opts=L_hdr=1000,L_sdr=100
```
![1 dd](https://github.com/user-attachments/assets/d3fc6ed0-b60a-4276-8e38-2e2e1f15c412)
![2 dd](https://github.com/user-attachments/assets/02f3d671-957c-4bdd-accd-10853a6c5400)

**Reply by stonix015** [reply]
Haha, i didnt want to stress you. All fine!

Just wanted to support you because i was already through that 1 month ago. (I am a fulltime scientist research nerd. 1 month on me is 2 years on another lol ...)

In the end, the st2094 mapping was way too overexposed to me. 

I can show you screenshots of two movies as an example of what i did. (Those movies we all legally own, you know for shure).
Note that even if you think so, nothing will be clipped out. It is just a black-level detection, it does not shift any information below zero.

**Feature Off:**
![prey_without](https://github.com/user-attachments/assets/703aa3d5-7bfc-4ea4-8003-1233b0dacc75)

**Feature On:**
![prey_with](https://github.com/user-attachments/assets/34dfd4eb-aa50-4376-819a-e2890c84553f)

And now for a movie with an adequate black level, you wont see a difference, nothing will be clipped:

**Feature Off:**
![bodies_without](https://github.com/user-attachments/assets/549ec427-4091-4263-a737-ba9f19375706)

**Feature_On:**
![bodies_with](https://github.com/user-attachments/assets/4837dbcd-8485-4ab6-a268-d9345910942c)


The thing is: 
This feature levels all movies to one picture quality. So you have way more possibilities to set mpv to your display. Without that feature i had to readjust everything for every film i watched.
So, you have to think of the advantages of this filter in an indirect way: If you set the target-contrast correctly and all details appear to you, there will be no movie that will gring you by its sudden greyish washed out colors! Think of that this way.....

**Comment by geextahslex** [comment]
Haha no, no stress but the last 2 weeks comparing all frames/scenes was a bit tedious. This is actually I think my 4th HDR tonemapping setup. 
1. Can you share that sample? You can cut only these few frames
2. I don't know how to compile so I would need a mpv-x86_64.exe for Windows 

Awesome to find someone who also experiments with that stuff, there are not many people around for that. Hopefully mpv can improve on that in the future. Because as you have seen I have to use an external tool for that.
You are right, now I always compare many productions (movies, series) to check if everything is fine. I used to finetune only for one production and then the next one looked completely messed up ^^

**Reply by stonix015** [reply]
Ahhhhhhhhhhhrrrggg shiet ..... i didnt know that youre working on windows ....

I even do not know how libplacebo is being involved in windows, what i know is that it IS involved .... 

I'll have a look.
I feel you .... This pic-comparing scene by scene is very frustrating, and even more regarding the bunch of possible combinations in the conf-setup.

**Reply by geextahslex** [reply]
Yapp, I don't even want to know how many possible combinatios there are between all of the settings. but to be fair hdr isn't that simple and also there are many "standards" and productions may vary. But this feature you have mentioned has made me curious 😈

**Comment by geextahslex** [comment]
@stonix015 okay I found this scene 00:21:12
this is your screenshot
![391601408-34dfd4eb-aa50-4376-819a-e2890c84553f](https://github.com/user-attachments/assets/5fc71e54-2991-46e9-9d3d-ffd092554404)

and this is mine without any patch
![fdfd](https://github.com/user-attachments/assets/a126a45d-4652-434b-8d60-65304f3fe51f)

**Reply by stonix015** [reply]
Okay, i tried something else:
![Bildschirmfoto_20241204_162340](https://github.com/user-attachments/assets/473e94a3-5b95-4a7e-857a-ad2454899738)

This screen is with settings as follows:
```
tone-mapping=linear (= all others, as long as no HDR peaks above 203nits are displayed)
target-peak=203 (= auto)
target-trc = bt.1886 (= auto)
target-contrast= 1000 (= auto) 
```
Warning: With the `bt.1886` trc curve, less `target-contrast` value means more contrast/darker. Setting `gamma2.2`, less value means less contrast/brighter/pushing up the dark areas to higher values regarding a display that really only has this contrast value.
-> The `bt.1886` curve seems to be wrong and causes your problems! `bt.1886` refers to a use of CRT displays from 30 years ago in a very dark room. Maybe it also can be used with an OLED nowadays, but your screen most likely has a gamma2.2 curve.
My thoughts on that: Even if this looks more contrasty, it is not "right", but only lacking of details in the dark areas.
Ironically Prey looks less worse with those settings, because it is as grey as it is. _(I call that grey, it's minimum PQ-value in this scene is 3,13%, refering to 0,02nits. This is far from being a black tone. On an OLED, this movie only uses a contrast of about 50'000:1 ... but this might look fine to those who own an OLED. )_

Ive been through that, this was where i started 2 months ago. The problem is that in a home theater with a projector  _(so all of the optics&contrast problems you got there)_ you wont see any details in dark areas. 
On well mixed dark horror movies that go down to 0,01% PQ or something like that, EVERYTHING will clip out with those settings. In Bodies-Bodies-Bodies my personal evolution was from "only seeing the half of her face" to "seeing her full face, hair, and some details in the background". _(Where the starting point was the settings as above)_

The trick on my filter is, that it is able to scale the movie down to a black level of zero, non-regarding of it's studio mix.
After that you can see every detail but without lacking of contrast and black-level.

Settings on my mpv, but WITH my filter, are:
```
tone-mapping = spline
target-peak=100
target-trc = gamma2.2
target-contrast= 2500 
```
The target-contrast=2500 has another meaning here: As i stated above, less value means pushing the dark areas of the image to higher values (more grey). But it also enables my filter to operate, as my filter only operates between 0 and the display black-level that results of the `target-contrast`. If i set `target-contrast = inf`, my filter does not do anything.
What i do is: I level my projector by its brightness lever to negative values to reach a full black in scenes that are completely dark.

**Reply by geextahslex** [reply]
Looks nice, I like it. I have no experience with projectors, I imagine that must be even harder to setup/finetune.
If you could show the exact code change maybe the guys could help out and compile a Windows version for me to test it.
I looked into the [thread](https://github.com/haasn/libplacebo/discussions/303) but this goes above my head ^^

**Reply by stonix015** [reply]
haasn has already answered my question: He is busy, but it might get happen!

Would be very cool, also for me as the version i use here has some dirty coding in it, and, worse, i deleted the avg-brightness feature as i needed the variable (ironically i didnt find where the variables are transferred, the only option left to me was using this variable that already exists)

Therefore the spline filter does not work anymore as expected, but also another idea of mine does not work, where i adapt to the avg image-brightness and calculate an absolute contrast due to indirect lightning room response problems with my projector setup. This feature could really be needed with all new adaptions, because the adaptive black-level feature pushes every scene to zero in its dark areas. In those areas, every detail will be lost on the projector setup.

The projector is an x7000 jvc. its a really pretty toy! It has so much capaibilities and contrast, that it can theoretically project the perfect picture. It just needs to get the right input (thats what i'm fighting for here). Due to its high panel-contrast, its picture also shows to be far more natural than with all concurrent projectors, so there is enough play for features like mine.

Greetings, Lukas

**Comment by kasper93** [comment]
Don't forget about https://mpv.io/manual/master/#options-tone-mapping-visualize to see what is the tone-mapping curve currently used.

**Comment by geextahslex** [comment]
@stonix015 Hi, are there any news regarding to these blacklevels?
